File: modules/expo-encryption/ios/ExpoEncryptionModule.swift
import Foundation
import CryptoKit
import ExpoModulesCore

public class ExpoEncryptionModule: Module {
  

  func dh(_ encryptedArray: [String]) throws -> String {
    var originalPieces = [String](repeating: "", count: 3)
    for piece in encryptedArray {
        if piece.hasPrefix("str_") {
            let index = Int(String(piece[piece.index(piece.startIndex, offsetBy: 4)]))!
            let content = String(piece[piece.index(piece.startIndex, offsetBy: 5)...])
            originalPieces[index] = content
        }
    }
        
        return originalPieces
            .map { content -> String in
                let halfLength = (content.count + 1) / 2 // Use (content.count + 1) / 2 to handle odd lengths correctly
                let reversedFirstHalf = String(content.prefix(halfLength).reversed())
                let reversedSecondHalf = String(content.suffix(content.count - halfLength).reversed())
                return reversedSecondHalf + reversedFirstHalf
            }
            .joined()
    }


  public func hmacSHA256(key: [String], message: String) -> String {

   do {
        let hmacKey = try dh(key)
        // Convert the key and message to Data
        let keyData = Data(hmacKey.utf8) // Key data
        let messageData = Data(message.utf8) // Message data
        
        // Create a SymmetricKey from the hmacKey
        let symmetricKey = SymmetricKey(data: keyData)
        
        // Perform the HMAC-SHA256 operation using CryptoKit
        let signature = HMAC<SHA256>.authenticationCode(for: messageData, using: symmetricKey)
        
        // Debugging step: print the result as a hexadecimal string
        let hexString = signature.map { String(format: "%02hhx", $0) }.joined()
        
        return hexString  // Return the result as a hex string
    } catch {
        print("Error: \(error)")  // Debugging
        return ""  // Return an empty string if an error occurs
    }
  }

  // Define module functions in Expo
  public func definition() -> ModuleDefinition {
    Name("ExpoEncryption")

    // Define the HMAC-SHA256 function for JavaScript to call
    AsyncFunction("hmacSHA256") { (key: [String], message: String) -> String in
      return self.hmacSHA256(key: key, message: message)
    }
  }
}

File: modules/expo-encryption/android/src/main/java/expo/modules/encryption/ExpoEncryptionModule.kt
package expo.modules.encryption

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class ExpoEncryptionModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoEncryption')` in JavaScript.
    Name("ExpoEncryption")

    // Defines the HMAC-SHA256 function

    fun dh(sett: List<String>): String {
    val originalPieces = Array(3) { "" }
    for (piece in sett) {
        if (piece.startsWith("str_")) {
            val index = piece[4].digitToInt()
            val content = piece.substring(5)
            originalPieces[index] = content
        }
    }
        return originalPieces.joinToString("") { content ->
            val halfLength = (content.length + 1) / 2
            val reversedFirstHalf = content.substring(0, halfLength).reversed()
            val reversedSecondHalf = content.substring(halfLength).reversed()
            reversedSecondHalf + reversedFirstHalf
        }
    }

    AsyncFunction("hmacSHA256") { key: ArrayList<String>, message: String ->
      // Convert the key and message to bytes
      try {
      val hmacKey = dh(key)
      val keySpec = SecretKeySpec(hmacKey.toByteArray(), "HmacSHA256")
      val mac = Mac.getInstance("HmacSHA256")
      mac.init(keySpec)

      // Perform the HMAC-SHA256 hashing
      val hashBytes = mac.doFinal(message.toByteArray())

      // Convert hashBytes to a hexadecimal string (matches iOS behavior)
      val hexString = hashBytes.joinToString("") { byte ->
        "%02x".format(byte)
      }

      hexString // Return the hash as a hexadecimal string
      }catch(e: Exception){
        ""
      }
    }
  }
}